### PR TITLE
rewrite the predecessors code to create a reduced graph

### LINKED ISSUE #39494: Refactor dependency graph reduction to allow for more work-products
**Background:** See the [general plan for skipping typeck](https://gist.github.com/nikomatsakis/0e225faaa35d5762e96d289cdfe1a4af).

The current code will not scale to saving more results than just the work-products. 

Pending PR:  https://github.com/rust-lang/rust/pull/39424

### PR DESCRIPTION
**Pull request by nikomatsakis**
The old code created a flat listing of "HIR -> WorkProduct" edges.
While perfectly general, this could lead to a lot of repetition if the
same HIR nodes affect many work-products. This is set to be a problem
when we start to skip typeck, since we will be adding a lot more
"work-product"-like nodes.

The newer code uses an alternative strategy: it "reduces" the graph
instead. Basically we walk the dep-graph and convert it to a DAG, where
we only keep intermediate nodes if they are used by multiple
work-products.

This DAG does not contain the same set of nodes as the original graph,
but it is guaranteed that (a) every output node is included in the graph
and (b) the set of input nodes that can reach each output node is
unchanged.

(Input nodes are basically HIR nodes and foreign metadata; output nodes
are nodes that have assocaited state which we will persist to disk in
some way. These are assumed to be disjoint sets.)

r? @michaelwoerister 

Fixes #39494

### CODE REVIEW

**Comment by nikomatsakis** [comment]
@michaelwoerister I did some measurements of syntex-syntax. The rust master takes 3.6 seconds to encode the dep-graph. This branch takes 2.6 seconds. So this seems like a clear win overall.

**Review by michaelwoerister — COMMENTED** [review]
I did one pass over this and it looks very good! I want to take another, closer look at the implementation and tests for the graph reduction algorithm.

File: src/librustc_incremental/persist/preds/compress/dag_id.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+use rustc_data_structures::graph::NodeIndex;
+use rustc_data_structures::unify::UnifyKey;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct DagId {
+    index: u32,
+}
+
+impl DagId {
+    pub fn from_in_index(n: NodeIndex) -> Self {
```
Comment (nikomatsakis) [inline src/librustc_incremental/persist/preds/compress/dag_id.rs]: Hmm, I meant `in` as in "the index in the INPUT graph". Perhaps `input_index`?

File: src/librustc_incremental/persist/preds/compress/test.rs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        "n3 -> n5",
+        "n3 -> n7",
+        "n5 -> n7",
+        "n5 -> n8",
+        "n6 -> n8",
+        "n6 -> n9"
+    ]);
+}
+
+//#[test]
```
Comment (nikomatsakis) [inline src/librustc_incremental/persist/preds/compress/test.rs]: Ah yes I forgot about that one.

File: src/librustc_incremental/persist/load.rs [hunk 3]
```diff
@@ -277,21 +262,37 @@ fn dirty_nodes<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
...
 }
 
+fn transitive_dirty_nodes(edges: &[SerializedEdge],
+                          mut dirty_nodes: DirtyNodes)
+                          -> DirtyNodes
+{
+    let mut len = 0;
+    while len != dirty_nodes.len() {
+        len = dirty_nodes.len();
+        for edge in edges {
```
Comment (nikomatsakis) [inline src/librustc_incremental/persist/load.rs]: Yeah, I was lazy. :) I can rewrite it to use a work-list or some such thing. One thing is that we don't have the edges indexed by their target, which we would want here. Perhaps I'll change how things are serialized to be a `Map<Target, Vec<Source>>`. I think we even build one of those in the "reduction" algorithm, so perhaps I should just return that (i.e., don't return a `Graph`, but some kind of `ReducedGraph`). I have to look at how the code works there.

**Comment by nikomatsakis** [comment]
@michaelwoerister I addressed (I think) your feedback w/ exception of commented out test, jfyi.

**Comment by nikomatsakis** [comment]
@michaelwoerister ok, ported the test.

File: src/librustc_incremental/persist/preds/compress/classify/test.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        IN -> B,
+        IN -> A,
+        A -> OUT,
+    };
+    let inputs = ["IN"];
+    let outputs = ["OUT"];
+    let mut reduce = GraphReduce::new(&graph, |n| inputs.contains(n), |n| outputs.contains(n));
+    Classify::new(&mut reduce).walk();
+
+    assert!(reduce.in_cycle(nodes("B"), nodes("C")));
```
Comment (michaelwoerister) [inline src/librustc_incremental/persist/preds/compress/classify/test.rs]: Can you also add `assert!(reduce.in_cycle(nodes("A"), nodes("C")));` and `assert!(reduce.in_cycle(nodes("A"), nodes("B")));`?

**Comment by michaelwoerister** [comment]
OK, I did another pass. Looks good to me. The graph reduction algorithm is very nice!
Please add the assertions in the test case, if you don't mind. Also, `make tidy` complains. Otherwise, consider the PR approved.

**Pull request merged**